helm-v3/005-test.md

```diff
@@ -0,0 +1,44 @@
+# Test Command
+
+## Overview
+
+`helm test` gives a chart author the ability to define actions on a helm release which validate that the chart is running as intended. This is useful for ensuring that a `helm install` or `helm upgrade` created/modified Kubernetes resources successfully and that the consequential `helm release` is functional.
+* A **test** is defined as a command and some arguments executed from a container running in a Kubernetes Pod. If the command completes inside the container with an exit code of 0, it is successful. All other exit codes indicate the test failed.
+* Tests must be defined in the `templates/tests/` directory. This used to be recommended as a best practice. See motivations for change below:
+  * Makes it much easier for a human to identify which templates are tests vs. which ones are manifests that define the application
+  * There is currently no built-in way to create supporting resources to use for running a test. Having a test directory will allow us to more easily have a way to organize and manage resources directly related to testing.
+  * This makes initial rendering and filtering resource manifests in the rendering engine much simpler to implement.
+However, making this a requirement will make the logic for rendering resources much easier to
+* Test annotations
+  * Each test pod must have an annotation to denote the expected condition.
+    * `helm.sh/test-expect-success: <true/false>`
+      * `helm.sh/test-expect-success: true` is used to 1) denote that the pod the annotation is applied to is a test and not a resource to be created during `helm install` and 2) denote the expected outcome of the test.
+    * In helm 2.x, tests used the annotation `helm.sh/hook: <test-success/test-fail>` to denote the expected condition. The test feature was implemented using the hooks system in helm 2 to add functionality in a backwards compatible way. We can move away from that now and make tests their own abstraction in helm v3 instead of re-using lifecycle hooks. This will make the test functionality more robust and will set a foundation for us to make the testing functionality even more powerful with possibily its own storage system for saving and fetching test results. It can also lay the foundation to allow us to more cleanly run tests as part of lifecycle hooks (like as a `post-install` hook).
```

Comment:
> It can also lay the foundation to allow us to more cleanly run tests as part of lifecycle hooks (like as a `post-install` hook).

It seems like we already have a pretty clean way of doing this in helm 2; if you want a test to run after an install (and cause helm to exit non-zero) and also to run with `helm test <release>`, then it is possible to define a pod with multiple hooks:

```yaml
kind: Pod
metadata:
  name: "{{.Release.Name}}-test"
  annotations:
    "helm.sh/hook": "post-install,test-success"
```

What would the migration path for this use case look like in helm 3?